Add tests and implementation for `extra-labels`

nautofon opened the pull request:
@truckermudgeon

This'll be the 2nd step in the review of truckermudgeon#47.

nautofon committed:
Tests and implementation for the extra-labels module

truckermudgeon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -64,6 +74,20 @@ export class LabelProducer {
+      target => new labelClass(target, this.dataProvider),
+    );
+
+    if (this.hasMeta) {
```
🤔 should `hasMeta` be a property of `LabelDataProvider`?

truckermudgeon replied:
WDYT of something like:

```ts
const mileageTargetLabels = dataProvider.mileageTargets.map(target => {
  const label = new labelClass(target, dataProvider);
  dataProvider.maybeAssignMeta(label);
  return label;
});
const nonMileageTargetLabels = dataProvider.missingLabels(mileageTargetLabels);
return [
    ...mileageTargetLabels,
    ...nonMileageTargetLabels,
];
```

nautofon replied:
We *could* move the `hasMeta` condition to `LabelDataProvider`, but I'm not sure that'd be an improvement.

Your proposal removes `hasMeta` entirely. `hasMeta` is just a performance optimization—arguably a premature one, seeing as the speed gain seems to be almost negligible for the current dataset, with something like 10% of execution time in my (very rough) measurement. But as implemented now, `hasMeta` doesn't really hurt, either.

I'd vote for keeping it as-is, but I don't have strong feelings one way or the other.

truckermudgeon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -337,6 +633,22 @@ export class LabelDataProvider {
   missingLabels(existing: Label[]): Label[] {
+    const existingByToken = new Map(
+      existing.map(label => [label.meta.token, label]),
+    );
```
Since the Map values are never used, this can be a Set of tokens.

nautofon replied:
Ahyes. Forgot that `Set` exists. 🤦

truckermudgeon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -337,6 +633,22 @@ export class LabelDataProvider {
+    );
+    return missingMeta.map(meta => {
+      const label = new GenericLabel(this);
+      Object.assign(label.meta, meta);
```
Use `.assignMeta()`?

truckermudgeon replied:
Ah, misread what was going on 🤦

truckermudgeon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -353,6 +665,14 @@ export class LabelDataProvider {
    *     `country` was given.
    */
   cityFromName(name: string, country: Country | undefined): City | undefined {
+    return Array.from(this.gameData.cities.values()).find(
```
I think it's [possible](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/find) to call `.values().find`, without first wrapping `.values()` in an `Array.from`.

nautofon replied:
You're right. Thanks!

truckermudgeon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -382,6 +706,13 @@ export class LabelDataProvider {
    * @see {@link clis/generator/geo-json/populated-places!ets2IsoA2}
    */
   countryFromCode(code: string): Country | undefined {
+    return Array.from(this.gameData.countries.values()).find(
```
Ditto.

nautofon committed:
Add name variants to unnamed TargetLabels

In a few cases, mileage targets excluded as "unnamed" come with a name or other useful hint in a variant name (e.g. Topaz Lake NV). Having this information readily available can make the manual editing of metadata easier.

For other labels than "unnamed" ones, the name variants only differ in spelling and alphabet, which makes them less useful. At the same time, the risk of interfering with metadata is higher for other labels.

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -133,10 +161,41 @@ export interface Label {
+    this.meta = {};
+  }
+
+  get isValid(): boolean {
+    return (
+      this.meta.kind != 'unnamed' &&
+      this.data.hasKnownCountryCode(this.meta) &&
+      this.meta.text != null &&
+      this.meta.easting != null &&
+      this.meta.southing != null
+    );
+  }
```
When grepping the project to see where I would need to change `isValid()` to `isValid` (as we discussed), I came across ["isValid()" in ScsArchive](https://github.com/truckermudgeon/maps/blob/dcb58729a39a72fdc89496c47944e35b8ab450fd/packages/clis/parser/game-files/scs-archive.ts#L158-L164). That was interesting for two reasons:

* You did string comparisons in there by checking strict equality `===`. Do you have a strong preference for *always* using strict equality where it's possible? (I tend to go the other way, using loose equality `==` where possible.) Should I update this branch to favor strict equality?

* That other method looks similar to this one, a bit simpler even. But it's not implemented as property getter. I still have no clear opinion over when to use and when not to use getters, and I'm not saying these two need to be the same. Still, it's an interesting data point I guess.

truckermudgeon replied:
* I do have a strong preference for strict equality where it's possible. Because JS makes me sad, and even with typings on variables catching most errors, loose equality checks always make me think of:

  <img width="102" height="60" alt="image" src="https://github.com/user-attachments/assets/9dee10af-dae7-4882-9a1e-1e473fdaf128" />

  and the less thinking I need to do of such things when reading code, the better.

* >  I still have no clear opinion over when to use and when not to use getters, and I'm not saying these two need to be the same. 

  Good catch! I probably should have written that as a prop getter. But I have mellowed out in the past 5 months w.r.t. fields / property getters / functions. I think as-written, this `isValid` prop getter is fine.

nautofon replied:
Okay, I tried to largely avoid implicit type conversions in conditions.

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -133,10 +161,41 @@ export interface Label {
+  toGeoJsonFeature(): GeoJSON.Feature<GeoJSON.Point, LabelMeta> {
+    const { easting, southing, ...meta } = this.meta;
+    const position = [easting, southing];
+    if (!position.every(v => v != null)) {
+      throw new ReferenceError('toGeoJsonFeature(): coordinates not defined');
+    }
```
I tried replacing that with this:
```ts
Preconditions.checkState(
  position.every(v => v != null),
  'toGeoJsonFeature(): coordinates not defined',
);
```
The [`checkState()` method is declared](https://github.com/truckermudgeon/maps/blob/dcb58729a39a72fdc89496c47944e35b8ab450fd/packages/libs/base/precon.ts#L19-L26) as `asserts condition is true`, which I expected would cause the complier to narrow the type just like `if (...) throw` does, but that didn't seem to work. Do you know why?

I also see I forgot to change `ReferenceError` to `Error` after trying `Preconditions.checkState()`. Will do that next time I update this branch.

truckermudgeon replied:
>  Do you know why?

Did some [playing around](https://www.typescriptlang.org/play/?ts=5.8.3#code/C4TwDgpgBAYg9nKBeKBvKpIC4oHIAWEANkXLgDRQBmCOBxpuAvlAD5obgR0DmCAJgCMQEClEEBDAE68Bw0UwBQi1IqjUArgDsAxsACWcLVAkBnUxCnAAFDqP99BozkEIiECVsoBbUzwD8OKbAUvpaPACUOGYWVqZQdloOTsaq6ur6VFDWAISJyYZaEWhq6erA+FJwAO5QWhC1AKJSVVLWvpEA3KXqSr2licEmOPCIKKgsZiZaIKUxljYSAHSY0EjreIQkZBEDRkP4AIx0W4zIJks0iAD01xiEJuYLhSauAG7QWtJV1fEABhI-o8oBAAB6QPQQfiKJTKNKaXQpR6xGz5RyFFxuDxeKAdQJQYKhcJRZELeJopH6eIhDTQeEZLK5CmFYr09IVH51BpQZqtdp+CLdMp9KAiwbAYawBDnCbAzyzdTzKzWZarZAbejbXC7dTiqBHE4MMjnZZXKC3e7QJVIiTvaD8OAQUxaXASrDWGGKIA).

The Precon function return types should be `asserts condition`, and not `asserts condition is true`. Apparently the latter loses information 😕 . Could you fix them in this PR? That should make `checkState` behave the way we want.

nautofon replied:
Found the [docs for assertion functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions). (I don't care much for the way they hide documentation in the release notes of old versions.)

The behavior of `asserts` observed here matches the documentation: The short form (just `asserts condition`) ensures the given boolean condition evaluates true for the rest of the scope, which can narrow the type of variables used within the condition.

On the other hand, the long form (`asserts val is ...`) only narrows the type of the given variable `val` to the defined type. The type of *other* variables used to create `val` is *not* narrowed.

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -166,8 +225,129 @@ export abstract class TargetLabel extends GenericLabel {
+      analysis.tidyName,
+    ]
+      .map(name => this.data.cityFromName(name, analysis.country))
+      .find(city => !!city);
```
Do you have a stylistic preference for filtering undefined objects?
- `.find(obj => obj)`
- `.find(obj => !!obj)`
- `.find(obj => obj != null)`

truckermudgeon replied:
Option 3.

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -166,8 +225,129 @@ export abstract class TargetLabel extends GenericLabel {
+    this.meta.easting = this.target.x;
+    this.meta.southing = this.target.y;
+
+    if (analysis.country) {
```
Do you have a stylistic preference for branching on a defined object?
- `if (obj)`
- `if (!!obj)`
- `if (obj != null)`

truckermudgeon replied:
Also option 3

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -187,6 +437,37 @@ export class AtsLabel extends TargetLabel {
+  protected override determineCity(analysis: TargetAnalysis): void {
+    super.determineCity(analysis);
+
+    // In ETS2, the mileage target editor name of cities is usually exactly
+    // the city token.
+    const cityByEditorName = this.data.cityFromToken(
+      this.target.editorName,
+      analysis.country,
+    );
```
JFTR:

The editor_name rule was meant to be a fall-back, but turned out to be important for ETS2 because the parser currently only provides English localized names for cities. As a result, the regular name rule (in the abstract superclass) often doesn't match in Europe. But almost all ETS2 cities are still successfully matched on the editor_name, so this might not matter much.

If the labels produced by this module are ever used with ETS2 for real, a determination would have to be made whether to continue with the map showing English exonyms, or to convert the map to display either endonyms or labels and UI in a user-defined language instead. Any decision for how to proceed regarding the matching of ETS2 city names to mileage targets can be postponed until then.

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -308,6 +594,20 @@ export class LabelDataProvider {
+    // the token, thus we don't filter on an `undefined` region check result.
+    this.metas = metas.filter(meta => this.isInRegion(meta) !== false);
+
+    this.metaByToken = new Map(this.metas.map(x => [x.token ?? '', x]));
+    this.metaByToken.delete('');
```
This feels more Perl-y than JavaScript-y to me, but maybe just because I haven't gotten used to modern JavaScript having stolen the `??` operator from Perl. 😅

I didn't manage to come up with something that's both clearer and similarly concise. Is it good enough to keep as-is?

truckermudgeon replied:
This is good enough as-is 👍

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -410,13 +754,18 @@ export class LabelDataProvider {
    * - `undefined` if the metadata record doesn't identify a country.
    */
   isInRegion(meta: LabelMeta): boolean | undefined {
+    if (!meta.country) {
```
Do you have a stylistic preference for branching on an undefined or empty string?
- `if(!str)`
- `if(str == null || str == '')`

truckermudgeon replied:
I don't think I do... I have a feeling I've done both in the codebase. FWIW, I like the latter a little more because I don't have to think too much about falsyness coercion when reading.

nautofon replied:
I suppose the latter is now more consistent with the style of other conditions.

nautofon commented on packages/clis/generator/geo-json/extra-labels.ts:
```diff
@@ -410,13 +754,18 @@ export class LabelDataProvider {
    * The name of the region for which game data is provided
    * (`'europe'` or `'usa'`).
    */
   get region(): RegionName {
+    return this.gameData.map;
   }
```
[Earlier](https://github.com/truckermudgeon/maps/pull/47#discussion_r1960669408), you asked if `region` could be a field in this class. I think it's easier to just access the property in `gameData` this way. What do you think?

truckermudgeon replied:
If `gameData.map` never changes (and looking at the type, I see I forgot to mark `map` as `readonly` 😬), then `region` could be implemented as a readonly field... something like:

```ts
export class Ets2Label extends TargetLabel {
  /** The name of the region... */
  readonly region: RegionName;
  constructor(...) {
    this.gameData = ...;
    this.region = this.gameData.map;
  }
```

But... I don't have a strong opinion. Happy either way.

nautofon replied:
In that case, I'll keep `region` as it is. Copying it in the constructor would effectively create a cache of `gameData.map`, and I see no reason to do that.

truckermudgeon reviewed (commented):
Blowing the dust off this PR 😅. Apologies for the delay!

nautofon commented:
Thanks for the review! 🙌 It's gonna take me a couple of days to get back to this though.

nautofon committed:
Code style: Avoid implicit type conversions in conditions

The only remaining cases of implicit conversion are null checks using loose equality. These are kept in order to avoid having to check for undefined as well as null.

nautofon committed:
Fix Preconditions.checkState() assertion

According to the TS docs, the form `asserts condition is ...` only narrows the type of the given variable `condition` to the defined type. The type of other variables used to create `condition` is not narrowed.

This kind of assertion needs to use the short form instead (without `is ...`). This ensures that the given boolean condition evaluates true for the rest of the scope, which can narrow the type of variables used within the condition.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

nautofon committed:
Replace null check with Preconditions.checkState()

nautofon committed:
Remove superfluous Array.from()

nautofon committed:
Replace Map (used as set) with Set

truckermudgeon reviewed (approved):
:shipit:
